File: dart/04-control-flow/loops.dart
void main() {
  // Counting for loop
  for (int i = 1; i <= 3; i++) {
    print('Count: $i');
  }

  // for-in loop over a list
  var fruits = ['apple', 'banana', 'cherry'];
  for (var fruit in fruits) {
    print('Fruit: $fruit');
  }

  // while loop
  int countdown = 3;
  while (countdown > 0) {
    print('T-minus $countdown');
    countdown--;
  }

  // do-while always runs the body at least once
  int n = 0;
  do {
    print('Runs once even though n is $n');
  } while (n > 0);
}

File: kotlin/04-control-flow/Loops.kt
fun main() {
    // for over an inclusive range
    print("Countup: ")
    for (i in 1..5) {
        print("$i ")
    }
    println()

    // downTo counts backwards, step changes the increment
    print("Countdown: ")
    for (i in 10 downTo 0 step 2) {
        print("$i ")
    }
    println()

    // until creates a half-open range (excludes the upper bound)
    print("Indices: ")
    for (i in 0 until 4) {
        print("$i ")
    }
    println()

    // iterating a collection with both index and value
    val fruits = listOf("apple", "banana", "cherry")
    for ((index, fruit) in fruits.withIndex()) {
        println("$index: $fruit")
    }

    // while loop runs while the condition is true
    var count = 3
    while (count > 0) {
        println("T-minus $count")
        count--
    }
    println("Liftoff!")

    // do/while runs the body at least once before checking
    var attempts = 0
    do {
        attempts++
        println("Attempt $attempts")
    } while (attempts < 2)
}
